x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java

```diff
@@ -0,0 +1,225 @@
+                }
+            };
+        } else {
+            return super.createGroup(replicas, settings);
+        }
+    }
+
+    protected IndexMetaData buildIndexMetaData(int replicas, Settings indexSettings, Map<String, String> mappings) throws IOException {
+        Settings settings = Settings.builder().put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
+            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, replicas)
+            .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+            .put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true)
+            .put(IndexSettings.INDEX_SOFT_DELETES_RETENTION_OPERATIONS_SETTING.getKey(), 10000)
+            .put(indexSettings)
+            .build();
+        IndexMetaData.Builder metaData = IndexMetaData.builder(index.getName()).settings(settings);
+        for (Map.Entry<String, String> typeMapping : mappings.entrySet()) {
+            metaData.putMapping(typeMapping.getKey(), typeMapping.getValue());
+        }
+        return metaData.build();
+    }
+
+    private ReplicationGroup createFollowGroup(int replicas) throws IOException {
+        Settings.Builder settingsBuilder = Settings.builder();
+        settingsBuilder.put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
+        return createGroup(replicas, settingsBuilder.build());
+    }
+
+    private ShardFollowNodeTask createShardFollowTask(ReplicationGroup leaderGroup, ReplicationGroup followerGroup) {
+        ShardFollowTask params = new ShardFollowTask(null, new ShardId("follow_index", "", 0),
+            new ShardId("leader_index", "", 0), 1024, 1, Long.MAX_VALUE, 1, 10240,
+            TimeValue.timeValueMillis(10), TimeValue.timeValueMillis(10), Collections.emptyMap());
+
+        BiConsumer<TimeValue, Runnable> scheduler = (delay, task) -> {
+            try {
+                Thread.sleep(delay.millis());
+            } catch (InterruptedException e) {
+                throw new AssertionError(e);
+            }
+            task.run();
+        };
+        AtomicBoolean stopped = new AtomicBoolean(false);
+        return new ShardFollowNodeTask(1L, "type", ShardFollowTask.NAME, "description", null, Collections.emptyMap(), params, scheduler) {
+            @Override
+            protected void updateMapping(LongConsumer handler) {
+                // noop, as mapping updates are not tested
+                handler.accept(1L);
+            }
+
+            @Override
+            protected void innerSendBulkShardOperationsRequest(List<Translog.Operation> operations, LongConsumer handler,
+                                                               Consumer<Exception> errorHandler) {
+                Runnable task = () -> {
+                    try {
+                        IndexShard primaryShard = followerGroup.getPrimary();
+                        TransportWriteAction.WritePrimaryResult<BulkShardOperationsRequest, BulkShardOperationsResponse> result =
```

Comment:
I think we should not execute these writes directly here but extend ESIndexLevelReplicationTestCase#ReplicationAction then run them via the infra of the new action (see ESIndexLevelReplicationTestCase#IndexingAction).

Reply:
So make `ReplicationAction` in `ESIndexLevelReplicationTestCase` have protected visibility, have a `CCRAction` in this test and then here do something like `new CCRAction(...).execute();` ?

Reply:
Yes, are you ok with this suggestion?

Reply:
Yes, just double checking whether I understood it correctly.

Reply:
added: e9125b51f7592d44e9d6cd1142e27d4d117303fc